kep/108-Memory-bandwidth-and-LLC-aware-filtering-and-scoring/README.md

```diff
@@ -0,0 +1,145 @@
+      - [Filter](#filter)
+      - [PreScore](#prescore)
+      - [Score](#score)
+  - [Known Limitations](#known-limitations)
+  - [Alternatives considered](#alternatives-considered)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Testing Plan](#testing-plan)
+  - [Implementation History](#implementation-history)
+  - [References](#references)
+<!-- /toc -->
+
+## Motivation
+
+Memory bandwidth and LLC (last level cache) are very important resources in a system. Kuberenetes scheduler hasn't taken them into consideration yet. But memory bandwidth and LLC contention may impact system performance very much. Furtherly, it causes bad SLA. Intel Resource Director Technology (RDT) is one example of the technology that can be leveraged to mitigate the impact of memory bandwidth and LLC contention. It is very helpful to ensure system performance stability and improve SLA. 
+
+## Goals
+1. Use scheduler plugin, which is the most Kubernetes native way, to implement memory bandwidth and LLC aware filtering and scoring.
+2. Leverage memory bandwidth and LLC relative metrics for filtering and scoring.
+3. Provide configurable weigths for prioritizing the metrics used in the scoring calculations.
+
+## Non-Goals
+
+## Use Cases
+
+## Terms
+LLC: last level cache.
+RDT: Intel Resource Director Technology.
+SLA: Service Level Agreement.
+MPKI: Misses Per Kilo Instructions.
+
+## Proposal
+We introduce some metrics such as memory bandwidth utilization, LLC occupation etc. At schedule stage, we can leverage these metrics to assist filting and scoring. It is hard to determine how important a resource is and how much resource is required. We did off-line analysis to different scenario and get a series of coefficient of correlation and generic usage value. The coefficient is called affinity and the generic usage value is called profile. The affinities and profiles are provided by a yaml configuration file. Detail about how the affinity and profile are used is described in the following design details section. The average value of a period of time is more meaningful than immediate value for some metrics, for instance, LLC occupancy and CPU usage. Metrics data aren't obtained from agents (such as cAdvisor) directly, instead they are stored in Prometheus first and then are retrieved on demand.
+
+## Design Details
+The filtering and scoring plugin leverages the metrics retrieved from Prometheus to determine the node candidates for running a pod. The metrics considered by the plugin at the moment are:
+
+* For filtering out oversubscribe nodes: 
+
+    * free memory available (GB)
+    * free memory bandwidth (GB/s)
+
+* For scoring:
+
+    * memory bandwidth (GB/s)
+    * memory latency (ns)
+    * Last Level Cache - l3 - Utilization (Bytes)
+    * Last Level Cache contention -- (MPKI - misses per Kilo Instructions)
+    * CPU utilization (%)
+
+the detail about filtering and scoring will be described below. 
+
+. Filtering
```

Comment:
Is it possible to run a controller to taint the nodes properly, so that you can rely on general taint/tolerations semantics to achieve the same goal?